README.md
# python-popcon

Get Debian [popularity contest](https://popcon.debian.org/) values for given
packages.


## Installation

```bash
$ pip install popcon
```


## Usage

The usage of this module is easy:

```python
>>> import popcon
>>> popcon.packages(['reportbug-ng'])
{'reportbug-ng': 323}
>>> popcon.packages(['reportbug-ng', 'reportbug'])
{'reportbug-ng': 323, 'reportbug': 75065}
```

The raw data (vote, old, recent, no-file) is also available, the sum of the raw
numbers is the number of installations as reported by `popcon.package`.

```python
>>> popcon.packages_raw(['reportbug-ng', 'reportbug'])
{'reportbug-ng': Package(vote=50, old=187, recent=86, no_files=0),
 'reportbug': Package(vote=5279, old=59652, recent=10118, no_files=16)}
```

Behind the scene popcon will try to use cached information saved in a file in
the ~/.cache/popcon directory. If the relevant file is not available, or older
than `EXPIRY` seconds (default is 7 days) it will download fresh data and save
that.

The cached data will be kept in memory unless `KEEP_DATA` is set to False.


## Command Line Interface

```
popcon reportbug python
             PACKAGE VALUE
           reportbug 180558
              python 110709
```

it also supports a more verbose mode on the command line:

```
popcon --verbose reportbug python
             PACKAGE VOTE       OLD        RECENT     NO FILES
           reportbug 5929       149243     25315      71
              python 3985       103591     3066       67
```


CHANGELOG.md
# Changelog

## [3.0.3] - 2022-06-29

* bumped version to allow for source-only upload in debian

## [3.0.2] - 2022-06-26

* bumped version to allow to migrate to Debian/testing (no changes)

## [3.0.1] - 2022-05-30

* Improved tabular output, for long package names
* Added the sum for verbose output

## [3.0.0] - 2022-05-29

* Removed functions deprecated since 2.0.1
* bumped minimim Python version to 3.7
* Added Python 3.10 to test suite
* use HTTPS everywhere
* updated dev dependencies
* Improved output and added verbose output
* Use argparse for CLI
* test the CLI

## [2.0.4] - 2021-08-08

* updated build system
* updated debian's build system
* use github actions instead of travis
* bumped minimal Python version to 3.6

## [2.0.3] - 2020-02-11

* Fixed missing version bump

## [2.0.2] - 2020-02-11

* Fixed main script to use the new-style functions introduced in 2.0.1
* Import new-style functions in __init__.py

## [2.0.1] - 2019-11-17

* deprecated functions with packages as positional arguments, the new ones use
  list of packages instead
* removed useles requirements.txt
* added html report to coverage


## [2.0.0] - 2019-01-13

* Dropped Python2 support
* moved to pytest
* added Makefile, setup.cfg, flake8
* Fixed lots of pep8 errors

## [1.5.1] - 2018-02-17

* Fixed __main__ handling in __main__.py

## [1.5.0] - 2017-11-04

* Updated packaging (setup.py, etc)
* Added popcon CLI for direct queries from the command line
* Removed dependency from xdg
* Added Travis CI
* Added unit tests
* Added changelog, license, readme


tests/popcon.py
import unittest

import popcon


class TestPopcon(unittest.TestCase):

    def test_version(self):
        try:
            popcon.__version__
        except NameError:
            self.fail()


if __name__ == '__main__':
    unittest.main()


tests/test_cli.py
import popcon.__main__ as cli


def test_main_single_package():
    cli.main('python-debianbts')


def test_main_multi_packages():
    cli.main('python-debianbts python3')


def test_main_multi_packages_verbose():
    cli.main('-v python-debianbts python3')


tests/test_popcon.py
import popcon


def test_version():
    assert hasattr(popcon, '__version__')


def test_packages():
    res = popcon.packages(['reportbug', 'python-debianbts'])
    assert len(res) == 2


def test_packages_raw():
    res = popcon.packages_raw(['reportbug', 'python-debianbts'])
    assert len(res) == 2


def test_source_packages():
    res = popcon.source_packages(['reportbug', 'python-debianbts'])
    assert len(res) == 2


def test_source_packages_raw():
    res = popcon.source_packages_raw(['reportbug', 'python-debianbts'])
    assert len(res) == 2


popcon/__main__.py
import argparse

from popcon import packages, packages_raw


def main(args=None):
    """Main entrypoint for the CLI.

    Parameters
    ----------
    args list[str]
        optional parameters, used for testing

    """
    args = parse_args(args)
    pkg = args.package
    # respective withs for package and numbers
    pwidth = 20
    nwidth = 10
    if not args.verbose:
        results = packages(pkg)
        if not results:
            return
        # calculate pwidth based on maximum length of package name, but at
        # least 7 characters to accommodate for the "PACKAGE" string in the
        # header
        pwidth = max(7, max(len(i) for i in results.keys()))
        print(f'{"PACKAGE":>{pwidth}} {"SUM":<{nwidth}}')
        for package, number in results.items():
            print(f'{package:>{pwidth}} {number:<{nwidth}}')
    else:
        results = packages_raw(pkg)
        if not results:
            return
        # calculate pwidth based on maximum length of package name, but at
        # least 7 characters to accommodate for the "PACKAGE" string in the
        # header
        pwidth = max(7, max(len(i) for i in results.keys()))
        print(
            f'{"PACKAGE":>{pwidth}} {"SUM":<{nwidth}} '
            f'{"VOTE":<{nwidth}} {"OLD":<{nwidth}} '
            f'{"RECENT":<{nwidth}} {"NO FILES":<{nwidth}}'
        )
        for package, values in results.items():
            vote = values.vote
            old = values.old
            recent = values.recent
            no_files = values.no_files
            sum_ = sum([vote, old, recent, no_files])
            print(
                f'{package:>{pwidth}} {sum_:<{nwidth}} '
                f'{vote:<{nwidth}} {old:<{nwidth}} '
                f'{recent:<{nwidth}} {no_files:<{nwidth}}'
            )


def parse_args(args=None):
    """Parse command line arguments.

    Parameters
    ----------
    args list[str]
        optional parameters, used for testing

    Returns
    -------
    argparse.Namespace

    """
    parser = argparse.ArgumentParser()

    parser.add_argument(
        'package',
        nargs="+",
        help="the package name(s)",
    )

    parser.add_argument(
        '-v', '--verbose',
        help="more verbose package output",
        action="store_true",
    )

    return parser.parse_args(args)


if __name__ == '__main__':
    main()


popcon/popcon.py
#!/usr/bin/env python


"""Get Debian popcon values for given packages.

The usage of this module is easy:

    >>> import popcon
    >>> popcon.package('reportbug-ng')
    {'reportbug-ng': 323}
    >>> popcon.package('reportbug-ng', 'reportbug')
    {'reportbug-ng': 323, 'reportbug': 75065}

The raw data (vote, old, recent, no-file) is also available, the sum of
the raw numbers is the number of installations as reported by
`popcon.package`.

    >>> popcon.package_raw('reportbug-ng', 'reportbug')
    {'reportbug-ng': Package(vote=50, old=187, recent=86, no_files=0),
            'reportbug': Package(vote=5279, old=59652, recent=10118,
            no_files=16)}

Behind the scene popcon will try to use cached information saved in a
file in the ~/.cache/popcon directory. If the relevant file is not
available, or older than `EXPIRY` seconds (default is 7 days) it will
download fresh data and save that.

The cached data will be kept in memory unless `KEEP_DATA` is set to
False.

"""


import warnings
import time
from urllib.request import Request, urlopen
import gzip
import io
import tempfile
import pickle
import os
import collections
import logging


logger = logging.getLogger(__name__)


XDG_CACHE_HOME = os.environ.get('XDG_CACHE_HOME',
                                os.path.expandvars('$HOME/.cache'))

__author__ = 'Bastian Venthur <venthur@debian.org>'

Package = collections.namedtuple(
    "Package", ["vote", "old", "recent", "no_files"])


# week in seconds
EXPIRY = 60 * 60 * 24 * 7
KEEP_DATA = True
cached_data = {}
cached_timestamp = {}


def _fetch(url):
    """Fetch all popcon results and return unparsed data.

    Parameters
    ----------
    url : str
        the url of the gzipped popcon results

    Returns
    -------
    txt : str
        the uncompressed data

    """
    request = Request(url)
    response = urlopen(request)
    txt = response.read()
    response.close()
    txt = _decompress(txt)
    return txt


def _parse(results):
    """Parse all-popcon-results file.

    Parameters
    ----------
    results : str
        the results file

    Returns
    -------
    ans : dict
        package name -> `Package` namedtuple mapping, containing the
        package information

    """
    ans = dict()
    results = results.splitlines()
    for line in results:
        elems = line.split()
        if elems[0] != b"Package:":
            continue
        ans[elems[1]] = Package(*(int(i) for i in elems[2:]))
    return ans


def _parse_stats(results):
    """Parse "statistics" files.

    Parameters
    ----------
    results : str
        the results file

    Returns
    -------
    ans : dict
        package name -> `Package` namedtuple mapping, containing the
        package information

    """
    ans = dict()
    results = results.splitlines()
    for line in results:
        elems = line.split()
        try:
            int(elems[0])
            int(elems[2])  # e.g. skip pass the "not in sid" pseudo-package
            if elems[1] == b"Total":
                continue
        except Exception:
            continue
        ans[elems[1]] = Package(*(int(i) for i in elems[3:]))
    return ans


def _decompress(compressed):
    """Decompress a gzipped string.

    Parameters
    ----------
    compressed : str
        the compressed string

    Returns
    -------
    data : str
        the uncompressed string

    """
    gzippedstream = io.BytesIO(compressed)
    gzipper = gzip.GzipFile(fileobj=gzippedstream)
    data = gzipper.read()
    return data


def packages(package_list):
    """Return the number of installations.

    The return value is a dict where the keys are the packages and the
    values the number of installations. If a package was not found it is
    not in the dict.

    Parameters
    ----------
    package_list : list of strings
        the package names

    Returns
    -------
    ans : dict
        packagename -> number of installations mapping

    """
    raw = packages_raw(package_list)
    ans = dict()
    for pkg, values in list(raw.items()):
        ans[pkg] = sum(values)
    return ans


def source_packages(package_list):
    """Return the number of installations, for source packages.

    See `package` for the format of the returned data.

    At present, this is only an approximation that instead gives the
    maximum value, out of the number of installations of any binary
    package belonging to each source package.

    Parameters
    ----------
    package_list : list of strings
        the package names

    Returns
    -------
    ans : dict
        packagename -> number of installations mapping

    """
    raw = source_packages_raw(package_list)
    ans = dict()
    for pkg, values in list(raw.items()):
        ans[pkg] = sum(values)
    return ans


def packages_raw(package_list):
    """Return the raw popcon values for the given packages.

    The return value is a dict where the keys are the packages and the
    values a named tuple of integers: (vote, old, recent, no-files)

    * vote: number of people who use this package regulary
    * old: is the number of people who installed, but don't use this
      package regularly
    * recent: is the number of people who upgraded this package recently
    * no-files: is the number of people whose entry didn't contain
      enough information (atime and ctime were 0)

    Parameters
    ----------
    package_list : list of strings
        the package names

    Returns
    -------
    ans : dict

    """
    return _packages_raw_generic(
        "https://popcon.debian.org/all-popcon-results.txt.gz",
        _parse, "debian", package_list)


def source_packages_raw(package_list):
    """Return the raw popcon values for the given source packages.

    See `package_raw` for the format of the returned data.

    At present, this is only an approximation that instead gives the
    maximum value, out of the number of installations of any binary
    package belonging to each source package.

    Parameters
    ----------
    package_list : list of strings

    Returns
    -------
    ans : dict

    """
    return _packages_raw_generic(
        "https://popcon.debian.org/sourcemax/by_inst.gz",
        _parse_stats, "debian-sourcemax", package_list)


def _packages_raw_generic(url, parse, key, package_list):
    """The work mule

    Parameters
    ----------
    url : str
        the url to use
    parse : function
        the parser function
    key : str
        "debian-sourcemax" or "debian"
    package_list : list of strings
        the debian package names

    Returns
    -------
    ans : dict

    """
    global cached_data, cached_timestamp
    # implements BASEDIRSPEC
    # https://standards.freedesktop.org/basedir-spec/basedir-spec-0.6.html
    dumpfile = os.path.join(
        XDG_CACHE_HOME,
        'popcon',
        "%s.%s" % (key, pickle.format_version))

    earliest_possible_mtime = max(
        time.time() - EXPIRY,
        os.stat(__file__).st_mtime)

    if (key in cached_data
            and cached_timestamp.get(key, 0) <= earliest_possible_mtime):
        del cached_data[key]

    data = cached_data.get(key, None)
    if (data is None
            and os.path.exists(dumpfile)
            and os.stat(dumpfile).st_mtime > earliest_possible_mtime):
        try:
            with open(dumpfile, 'rb') as fh:
                data = pickle.load(fh)
            cached_timestamp[key] = os.stat(dumpfile).st_mtime
        except Exception:
            import traceback
            warnings.warn("Problems loading cache file: %s" % dumpfile)
            traceback.print_exc()

    if data is None:
        data = _fetch(url)
        data = parse(data)
        # i still think that makedirs should behave like mkdir -p
        if not os.path.isdir(os.path.dirname(dumpfile)):
            os.makedirs(
                os.path.dirname(dumpfile),
                mode=0o700)  # mode according to BASEDIRSPEC
        # as soon as python2.6 is in stable, we can use delete=False
        # here and replace the flush/rename/try:close sequence with the
        # cleaner close/rename.
        temp = tempfile.NamedTemporaryFile(dir=os.path.dirname(dumpfile))
        pickle.dump(data, temp)
        temp.flush()
        os.rename(temp.name, dumpfile)
        try:
            temp.close()
        except OSError:
            pass
        cached_timestamp[key] = time.time()
    ans = dict()
    for pkg in package_list:
        # Lookup using bytestrings, but always index results by the
        # original so that callsites can look it up.
        lookup = pkg if isinstance(pkg, bytes) else pkg.encode('utf-8')
        if lookup in data:
            ans[pkg] = data[lookup]
    if KEEP_DATA:
        cached_data[key] = data
    return ans
